snakemake_interface_storage_plugins/storage_provider.py

```diff
@@ -36,48 +51,91 @@ def __bool__(self):
+    valid queries for a specific storage provider.
+
+    Parameters
+    ----------
+    query : str
+        Example query string demonstrating correct format.
+    description : str
+        Human-readable explanation of what the query does.
+    type : QueryType
+        Whether this example is for input, output, or both.
+    """
+
     query: str
     description: str
     type: QueryType
 
 
+@dataclass
 class StorageProviderBase(ABC):
-    """This is an abstract class to be used to derive remote provider classes.
-    These might be used to hold common credentials,
-    and are then passed to StorageObjects.
+    """Abstract base class for Snakemake storage providers.
+
+    Defines the interface for interacting with external storage systems
+    like S3, GCS, HTTP, etc. Storage providers handle authentication,
+    rate limiting, caching, and mapping between remote resources and
+    local files within Snakemake workflows.
+
+    Parameters
+    ----------
+    local_prefix : Path
+        Directory where remote files are cached locally.
+    settings : StorageProviderSettingsBase, optional
+        Provider-specific configuration options.
+    keep_local : bool, default=False
+        Whether to retain local copies after workflow completion.
+    retrieve : bool, default=True
+        Whether to automatically fetch remote files when referenced.
+    is_default : bool, default=False
+        Whether this provider is the default for its protocol.
     """
 
-    def __init__(
-        self,
-        local_prefix: Path,
-        settings: Optional[StorageProviderSettingsBase] = None,
-        keep_local=False,
-        retrieve=True,
-        is_default=False,
-    ):
+    # Class attributes with type hints
+    local_prefix: Path
+    settings: Optional[StorageProviderSettingsBase] = None
+    keep_local: bool = False
+    retrieve: bool = True
+    is_default: bool = False
+    _rate_limiters: Dict[Any, Throttler] = field(default_factory=dict, init=False)
+
+    def __post_init__(self):
+        """Hook for subclasses to perform additional initialization.
+
+        Subclasses may override this method to perform additional setup
+        after the base class has been initialized and then call the base
+        class implementation using `super().__post_init__()`.
+        """
         try:
-            local_prefix.mkdir(parents=True, exist_ok=True)
+            self.local_prefix.mkdir(parents=True, exist_ok=True)
         except OSError as e:
             raise WorkflowError(
-                f"Failed to create local storage prefix {local_prefix}", e
+                f"Failed to create local storage prefix {self.local_prefix}", e
             )
-        self.local_prefix = local_prefix
-        self.settings = settings
-        self.keep_local = keep_local
-        self.retrieve = retrieve
-        self.is_default = is_default
-        self._rate_limiters = dict()
-        self.__post_init__()
```

Comment:
I remember now why I chose to use __init__ instead of a dataclass. This way, plugins that implement this class can start from scratch with __post_init__ and they cannot accidenally overwrite anything that is done in the __post_init__ of this base class.
I guess you should be able to achieve the same type annotations as with a dataclass by adding the types to the attributes that are defined in __init__.